pyresample/geometry.py

```diff
@@ -717,37 +717,282 @@ def _do_transform(src, dst, lons, lats, alt):
+            dst_lons = np.concatenate((dst_lons, extended_side2_lonlats[0]), axis=0)
+
+        # Extend on x direction (side1 and side3)
+        if x > 0:
+            list_side1 = (dst_lons[:, -2], dst_lats[:, -2], dst_lons[:, -1], dst_lats[:, -1])
+            list_side3 = (dst_lons[:, 1], dst_lats[:, 1], dst_lons[:, 0], dst_lats[:, 0])
+            extended_side1_lonlats = _get_extended_lonlats(*list_side1, npts=x, transpose=False)
+            extended_side3_lonlats = _get_extended_lonlats(*list_side3, npts=x, transpose=False)
+            dst_lats = np.concatenate((dst_lats, extended_side1_lonlats[1]), axis=1)
+            dst_lats = np.concatenate((extended_side3_lonlats[1][:, ::-1], dst_lats), axis=1)
+            dst_lons = np.concatenate((dst_lons, extended_side1_lonlats[0]), axis=1)
+            dst_lons = np.concatenate((extended_side3_lonlats[0][:, ::-1], dst_lons), axis=1)
+
+        # Back-conversion array as input format
+        lons, _ = _convert_2D_array(dst_lons, to=src_lons_format, dims=['y', 'x'])
+        lats, _ = _convert_2D_array(dst_lats, to=src_lats_format, dims=['y', 'x'])
+
+        # Add additional info if the source array is a DataArray
+        if isinstance(self.lats, xr.DataArray) and isinstance(self.lons, xr.DataArray):
+            lats.attrs = self.lats.attrs.copy()
+            lons.attrs = self.lons.attrs.copy()
+
+        # Return the extended SwathDefinition
+        return SwathDefinition(lons, lats)
+
+    def reduce(self, x=0, y=0):
+        """Reduce the swath definition along x (along-track) and y (across-track) dimensions.
+
+        By default, it does not reduce on any direction.
+        To reduce of n pixel on both sides of the across-track direction, call swath_def.reduce(x=0, y=2).
+        """
+        # Check input validity (ensure reduced area is at least 2x2)
+        height = self.lats.shape[0]
+        width = self.lats.shape[1]
+        x = int(x)
+        y = int(y)
+        if x < 0 or y < 0:
+            raise ValueError('x and y arguments must be positive integers.')
+        if x >= np.floor(width / 2):
+            max_x = int(np.floor(width / 2)) - 1
+            raise ValueError("""You can at maximum reduce the along-track direction (x)
+                             of SwathDef by {} pixels on each side.""".format(max_x))
+        if y >= np.floor(height / 2):
+            max_y = int(np.floor(height / 2)) - 1
+            raise ValueError("""You can at maximum reduce the across-track direction (y)
+                             of SwathDef by {} pixels on each side.""".format(max_y))
+
+        # Return SwathDefinition if nothing to reduce
+        if x == 0 and y == 0:
+            return self
+
+        # Return the reduced SwathDefinition
+        lats = self.lats[slice(0 + y, height - y), slice(0 + x, width - x)]
+        lons = self.lons[slice(0 + y, height - y), slice(0 + x, width - x)]
```

Comment:
I added it because is the opposite of `extend`. It appeared to me would have been nice having boths.
If one want to extend a swath, then upsample and then reduce it a bit, to me the code 
`swath_def.extend(y=20, x=0).upsample(x=5,y=5).reduce(y=2, x=0)` looks more readable than 
`swath_def.extend(y=20, x=0).upsample(x=5,y=5)[slice(2, -2),:]` .

Reply:
I beg to differ :smile: , especially since `reduce` means something totally different in functional programming terms ( https://docs.python.org/3/library/functools.html#functools.reduce ) and that I don't find it intuitive at all that the function shrinks the data on both sides at once. I can only imagine the number of WTFs per minute we will get when reading code using this in the future :joy: . But maybe it's just me? ping @djhoese @pnuu

Reply:
I agree with Martin on this (shocking, I know). I think the heavy existing usage of `reduce` in functional programming makes this an unfortunate naming. Dask has a `reduce` function, but it resembles the functional approach where you provide it functions to call on portions of the array. I think someone using this method would assume something similar or at least pre-built options for how the array should be reduced/aggregated (first pixel in a slice, last pixel, mean of the pixels, etc). I'm also not sure why the `height - ` is needed, wouldn't a negative index work just as well?

Reply:
I changed the name to `shrink`.
I used the `slice(..., height -y)` pattern instead of `slice(..., -y) `  to avoid obtaining `slice(..., -0)` when `x=0`.
I changed the input arguments to `left=0, right=0, bottom=0, top=0` to follow the new `extend` convention